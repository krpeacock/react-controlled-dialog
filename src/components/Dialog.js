import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";

const defaultCloseButton = ({ onClick }) => {
  return (
    <button onClick={onClick} aria-label="Close">
      x
    </button>
  );
};

const BaseDialog = React.forwardRef((props, ref) => (
  <dialog ref={ref} {...props} />
));

function isMouseEventInClientArea(event) {
  var element = event.currentTarget;
  var rect = element.getBoundingClientRect();
  var minX = rect.left + element.clientLeft;
  var x = event.clientX;
  if (x < minX || x >= minX + element.clientWidth) return false;
  var minY = rect.top + element.clientTop;
  var y = event.clientY;
  if (y < minY || y >= minY + element.clientHeight) return false;
  return true;
}

const Dialog = ({
  children,
  isOpen,
  setOpen,
  type = "dialog",
  CloseButton = defaultCloseButton,
  DialogComponent = BaseDialog,
  closeOnBackdropClick = false,
  onClick
}) => {
  const ref = React.createRef();
  let [init, setInit] = useState(false);

  useEffect(() => {
    if (!init) {
      // Register the dialog with dialog-polyfill once, if available
      if (window.dialogPolyfill) {
        window.dialogPolyfill.registerDialog(ref);
      }
      setInit(true);
    }
  });
  let dialog;
  useEffect(() => {
    dialog = ref.current || {};
    if (isOpen && !dialog.open) {
      if (type === "modal") {
        return dialog.showModal();
      }
      return dialog.show();
    } else if (!isOpen && dialog.open) {
      dialog.close();
    }
  });

  const handleBackdropClick = event => {
    function decideToClose() {
      if (!isMouseEventInClientArea(event)) {
        dialog.close();
      }
    }
    if (typeof closeOnBackdropClick === "function") {
      if (closeOnBackdropClick()) {
        decideToClose();
      } else if (closeOnBackdropClick) {
        decideToClose();
      }
    }
  };

  return (
    <DialogComponent
      ref={ref}
      role={type}
      onClick={event => {
        if (onClick) onClick(event);
        handleBackdropClick(event);
      }}
    >
      <CloseButton onClick={() => setOpen(false)} aria-label="Close" />
      {children}
    </DialogComponent>
  );
};

Dialog.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]),
  isOpen: PropTypes.bool.isRequired,
  setOpen: PropTypes.func.isRequired,
  type: PropTypes.oneOf(["dialog", "modal"]),
  closeButton: PropTypes.node,
  DialogComponent: PropTypes.node,
  closeOnBackdropClick: PropTypes.bool || PropTypes.func,
  onClick: PropTypes.func
};

export default Dialog;
