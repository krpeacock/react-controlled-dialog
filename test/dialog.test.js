import React from "react";
import Dialog from "../src/components/Dialog";
import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import {
  render,
  fireEvent,
  cleanup,
  waitForElement
} from "react-testing-library";

const show = jest.fn();
const showModal = jest.fn();

const defaultShowModal = HTMLDialogElement.prototype.showModal;
const defaultShow = HTMLDialogElement.prototype.show;

beforeEach(() => {
  HTMLDialogElement.prototype.showModal = showModal;
  HTMLDialogElement.prototype.show = show;
});
afterEach(() => {
  HTMLDialogElement.prototype.showModal = defaultShowModal;
  HTMLDialogElement.prototype.show = defaultShow;
  cleanup();
});

configure({ adapter: new Adapter() });

describe("Dialog tests", () => {
  it("should run without crashing", () => {
    // TODO - replace with shallow once Enzyme supports hooks
    const { container } = render(<Dialog isOpen={false} setOpen={() => {}} />);
    expect(container.querySelector("dialog")).toBeTruthy();
    expect(container).toMatchSnapshot();
  });
  it("should accept a custom Dialog component", () => {
    const CustomDialog = React.forwardRef((props, ref) => (
      <dialog id="custom" />
    ));

    const { container } = render(
      <Dialog
        isOpen={false}
        setOpen={() => {}}
        DialogComponent={CustomDialog}
      />
    );
    expect(container.querySelector("#custom")).toBeTruthy();
  });
  it("should open the dialog as a dialog if type is not provided", () => {
    const { container } = render(<Dialog isOpen={true} setOpen={jest.fn()} />);
    expect(HTMLDialogElement.prototype.showModal).toBeCalledTimes(0);
    expect(HTMLDialogElement.prototype.show).toBeCalled();
  });
  it("should open the dialog as a modal if type is modal", () => {
    const { container } = render(
      <Dialog isOpen={true} setOpen={jest.fn()} type="modal" />
    );
    expect(HTMLDialogElement.prototype.showModal).toBeCalled();
    expect(HTMLDialogElement.prototype.show).toBeCalledTimes(0);
  });
});
