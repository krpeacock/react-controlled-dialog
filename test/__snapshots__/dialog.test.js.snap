// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog tests should run without crashing 1`] = `
<div>
  <dialog
    role="dialog"
  >
    <button
      aria-label="Close"
    >
      x
    </button>
  </dialog>
</div>
`;
